### PR TITLE
Serial topics

### PR DESCRIPTION
**Pull request by mlecriva**

### CODE REVIEW

File: arduino/common/SerialTopics.cpp [hunk 1]
```diff
@@ -0,0 +1,98 @@
+#include "SerialTopics.h"
+#include "Clock.h"
+
+// Global instance
+
+SerialTopics topics;
+
+void SerialTopics::SUBSCRIBE(SerialTalks &talks, Deserializer &input, Serializer &output)
```
Comment (Enderdead) [inline arduino/common/SerialTopics.cpp]: Define two functions (SUBSCRIBE and UNSUBSCRIBE) is a little bit inefficient. You can instead, define a CHANGE_STATUS, and perform a switch case with an extra argument given by the python part.

File: arduino/common/SerialTopics.cpp [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        topics.getSubscriptions()[opcode].enable = false;
+        output.write<bool>(true);
+    }
+    else
+    {
+        output.write<bool>(false);
+    }
+}
+
+void SerialTopics::GET_CONTEXT(SerialTalks &talks, Deserializer &input, Serializer &output)
```
Comment (Enderdead) [inline arduino/common/SerialTopics.cpp]: I don't really get the need of this method.

File: arduino/common/SerialTopics.h [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+private:
+    SerialTalks *_talks;
+
+    subscription_t _subscriptions[SERIALTOPICS_MAX_OPCODE]; /*!< Listes des souscriptions enregistrées avec un OPCode associé.*/
+
+    static void SUBSCRIBE(SerialTalks &talks, Deserializer &input, Serializer &output);
+    static void UNSUBSCRIBE(SerialTalks &talks, Deserializer &input, Serializer &output);
+    static void GET_CONTEXT(SerialTalks &talks, Deserializer &input, Serializer &output);
+
+    static void DEFAULT_HANDLER(Serializer &output);
```
Comment (Enderdead) [inline arduino/common/SerialTopics.h]: I didn't understand what is used for ? Add some docstring ? ;-)

File: raspberrypi/common/serialtopics.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    def getTimestep(self):
+        return self.timestep
+
+    def getHandler(self):
+        return self._handler
+
+
+class SerialTopics:
+    def __init__(self, parent):
+        self.parent = parent
```
Comment (Enderdead) [inline raspberrypi/common/serialtopics.py]: Add a `isinstanceof`to be sure it's a SerialTalks instance

File: raspberrypi/common/serialtopics.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+        return bool(output.read(BYTE))
+
+    def unsubscribe(self, topic):
+        output = self.parent.execute(
+            UNSUBSCRIBE_OPCODE, BYTE(topic.getID()))
+
+        return bool(output.read(BYTE))
+
+    def get_context(self, topic):
```
Comment (Enderdead) [inline raspberrypi/common/serialtopics.py]: Very usefull ??

File: raspberrypi/common/serialtopics.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+TOPIC_2_OPCODE = 0X01
+TOPIC_3_OPCODE = 0X02
+TOPIC_4_OPCODE = 0X03
+TOPIC_5_OPCODE = 0X04
+
+TOPIC_DEFAULT_TIMESTEP = 100  # ms
+
+
+class Topic:
+    def __init__(self, id, timestep=TOPIC_DEFAULT_TIMESTEP):
```
Comment (Enderdead) [inline raspberrypi/common/serialtopics.py]: Handler class is a little bit over kill. You can instead add an arg list containing all deserializer for dealing  with incoming packets.

File: arduino/common/SerialTopics.h [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+#define SERIALTOPICS_DEFAULT_TIMING 100 //ms
+
+#define SERIALTOPICS_MAX_OPCODE 5
+
+#define SUBSCRIBE_OPCODE (SERIALTALKS_RESEVED_OPCODE_7)
+#define UNSUBSCRIBE_OPCODE (SERIALTALKS_RESEVED_OPCODE_8)
+#define GET_CONTEXT_OPCODE (SERIALTALKS_RESEVED_OPCODE_9)
+
+class SerialTopics
```
Comment (Enderdead) [inline arduino/common/SerialTopics.h]: It would be great if you inhearit from Periodic Process.

File: arduino/common/SerialTopics.cpp (hunk 2 above)
Comment (mlecriva) [reply arduino/common/SerialTopics.cpp]: It is useful now to check the topic state by the python part. I will remove this method on the final release

**Pull request merged**